File: leetcode/336. Palindrome Pairs/solution.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> palindromePairs(vector<string>& words) {
        vector<vector<int>> result;
        unordered_map<string, int> dict;
        int i, j, size = words.size();
        string left, right, tmp;
        for(i = 0; i < size; i++) {
            tmp = words[i];
            reverse(tmp.begin(), tmp.end());
            dict[tmp] = i;
        }
        
        for(i = 0; i < size; i++) {
            for(j = 0; j < words[i].size(); j++) {
                left = words[i].substr(0, j);
                right = words[i].substr(j);
                if(dict.find(left) != dict.end() && dict[left] != i && isPalindrome(right)) {
                    result.push_back({i, dict[left]});
                    if(left.empty())
                        result.push_back({dict[left], i});
                }
                if(dict.find(right) != dict.end() && dict[right] != i && isPalindrome(left))
                    result.push_back({dict[right], i});
            }
        }
        return result;
    }

private:
    bool isPalindrome(string s) {
        int i = 0;
        int j = s.size() - 1;
        while (i < j) {
            if (s[i++] != s[j--]) return false;
        }
        return true;
    }
};

File: leetcode/336. Palindrome Pairs/solution.py
class Solution:
    def palindromePairs(self, words: List[str]) -> List[List[int]]:
        is_palindrome = lambda x: x == x[::-1]
        res = []
        dic = {}
        n = len(words)
        for i in range(n):
            dic[words[i][::-1]] = i
        for i in range(n):
            for j in range(len(words[i])):
                left = words[i][:j]
                right = words[i][j:]
                if left in dic and dic[left] != i and is_palindrome(right):
                    res.append([i, dic[left]])
                    if not left:
                        res.append([dic[left], i])
                if right in dic and dic[right] != i and is_palindrome(left):
                    res.append([dic[right], i])
        return res

if __name__ == '__main__':
    s = Solution()

    # Example 1:

    # Input: ["abcd","dcba","lls","s","sssll"]
    # Output: [[0,1],[1,0],[3,2],[2,4]] 
    # Explanation: The palindromes are ["dcbaabcd","abcddcba","slls","llssssll"]
    
    # ex1 = ["abcd","dcba","lls","s","sssll"]
    # print(s.palindromePairs(ex1))

    # Example 2:

    # Input: ["bat","tab","cat"]
    # Output: [[0,1],[1,0]] 
    # Explanation: The palindromes are ["battab","tabbat"]
    # ex2 = ["bat", "tab", "cat"]
    # print(s.palindromePairs(ex2))

    # Example 3:

    ex3 = ["a",""]
    print(s.palindromePairs(ex3))
